compiler/cli/src/org/jetbrains/kotlin/cli/jvm/plugins/DependencyShadingPluginClassLoader.kt

```diff
@@ -0,0 +1,68 @@
+package org.jetbrains.kotlin.cli.jvm.plugins
+
+import org.jetbrains.org.objectweb.asm.ClassReader
+import org.jetbrains.org.objectweb.asm.ClassWriter
+import org.jetbrains.org.objectweb.asm.commons.ClassRemapper
+import org.jetbrains.org.objectweb.asm.commons.Remapper
+import java.io.IOException
+import java.io.InputStream
+import java.net.URL
+import java.net.URLClassLoader
+import java.security.AccessControlContext
+import java.security.AccessController
+import java.security.PrivilegedActionException
+import java.security.PrivilegedExceptionAction
+
+/**
+ * DependencyShadingPluginClassLoader will load a plugin jar, applying the necessary shading as required
+ * by the variant of the Kotlin Compiler which is consuming the plugin.
+ */
+class DependencyShadingPluginClassLoader(urls: Array<URL>, parent: ClassLoader) : URLClassLoader(urls, parent) {
+
+    val accessControlContext: AccessControlContext
+    val bytecodeLoader = URLClassLoader(urls, null)
+
+    init {
+        accessControlContext = AccessController.getContext()
+    }
+
+    // True if the Kotlin Compiler loading this plugin is shaded, and thus the plugin jar should also be shaded
+    val shouldBeShaded = com.intellij.mock.MockProject::class.java.name.startsWith("org.jetbrains.kotlin.")
+
+    @Throws(ClassNotFoundException::class)
+    override fun findClass(name: String): Class<*>? {
+        return try {
+            AccessController.doPrivileged(
+                PrivilegedExceptionAction {
+                    val path = name.replace('.', '/') + ".class"
+                    val originalBytes = bytecodeLoader.getResourceAsStream(path)?.readBytes()
+
+                    if(originalBytes == null) throw ClassNotFoundException(name)
+
+                    try {
+                        val cw = ClassWriter(ClassWriter.COMPUTE_MAXS)
+
+                        val shader = ClassRemapper(cw, object : Remapper() {
+                            override fun map(internalName: String): String {
+                                if(shouldBeShaded && internalName.startsWith("com/intellij/"))
```

Comment:
`com.intellij` is not the only prefix that is shaded, there's also at least `com.google.*`, `it.unimi.dsi.fastutil.*`, `org.jline.*`, `org.picocontainer.*`